include/kernel.h

```diff
@@ -0,0 +1,534 @@
+ * itself offers libraries that need threading to provide a easy-to-use
+ * API.
+ *
+ * A context-switch between threads can happen in the following situations:
+ *
+ *   * Preemptively: when a hardware interrupt generates a event that
+ *     wakes up a thread (see #kevent). This is the most common scenario:
+ *     an event connected to an interrupt is fired, and a higher priority
+ *     thread waiting for the event is woken up.
+ * 
+ *   * Cooperatively: when a thread explicitly calls a blocking function
+ *     like #kthread_sleep or #kthread_wait_event to wait for an event.
+ *     The kernel will put the current thread to sleep and schedule the
+ *     highest-priority thread among the ready ones, including threads
+ *     that have a priority lower than the one that goes to sleep. When
+ *     the event is fired, the lower-priority thread will be preempted
+ *     to give back control to the thread that just woke up.
+ *
+ *   * Cooperatively: when a thread explicitly calls #kthread_yield to
+ *     leave the CPU to other threads that might be ready at the same time.
+ *     Notice that, given the hard real-time guarantee, this only makes
+ *     sense if there are multiple ready threads at the same priority level,
+ *     as calling #kthread_yield will not schedule a lower-priority thread.
+ *     It is normally not necessary to call #thread_yield.
+ *
+ * In addition to threads (#kthread_t) and events (#kevent_t), the kernel
+ * library offers a message passing communication primitive called "message
+ * box" (#kmbox_t). A mailbox is a circular buffer of messages called
+ * "mails"; any number of threads can try to send mails into a mailbox,
+ * and any number of threads can try to read mails from a mailbox. Once a
+ * thread receives a mail from a mailbox, the mail is removed and other
+ * threads will not see it anymore. 
+ *
+ * Mailboxes can be useful as communicative primitives among threads
+ * (eg: to send tasks to other threads), or a synchronization primitive
+ * to wait for multiple events (see #kmbox_attach_event).
+ *
+ * @{
+ */
+
+/**
+ * @brief a kernel thread for parallel execution
+ *
+ * This structure represents a thread that can be scheduled
+ * for parallel execution. Create a thread with #kthread_new,
+ * allocating the required memory for the stack.
+ *
+ * Normally, there's no need to explicitly manage the lifetime
+ * or state of a thread. Once the thread is created, it is
+ * immediately started, and the kernel will scheduled it when
+ * required. If the thread exits (by simply returning by its
+ * entry point function), it will be cleaned up and the memory
+ * released.
+ *
+ * #kthread_new returns a pointer to the thread (a pointer
+ * to kthread_t) that can be used to externally manage the
+ * thread like changing its priority or killing it.
+ */
+typedef struct kthread_s
+{
+	/** Pointer to the top of the stack, which contains the thread register state */
```

Comment:
Maybe we should not call this stack to reduce confusion

Reply:
I was thinking about this name a few days ago. This variable is the one that basically contains the thread SP register *while the thread is sleeping*. It gets written when the scheduler puts the thread to sleep, and then it's read to restore the SP register when the thread is scheduled. Its type is `reg_block_t*` because we know how to the top of the stack looks like when the interrupt is triggered (just like when an exception is triggered).

Do you think `save_sp` would be better?

Also: for reducing future bugs, we might want to write NULL in this variable when the thread is scheduled. In fact, while the thread is running, this points to some point in the stack that contains nobody knows what (it used to contain the exception stack during exception, but now it contains random data).

Reply:
So the only thing we care is SP and we keep the struct to help with the offset from what I understand from your explanation. If that's the case, we may change it to a `uint64_t` and actually write the contents via `offsetof` etc.

Or if that's not something practical, what about `last_interrupt_frame` maybe?

Reply:
> So the only thing we care is SP and we keep the struct to help with the offset from what I understand from your explanation. If that's the case, we may change it to a uint64_t and actually write the contents via offsetof etc.

Uhm I'm not sure I follow. A `thread_t` is a thread descriptor. For a sleeping thread, we need to keep its stack pointer saved in the descriptor, otherwise we have no way to schedule it again. We could keep the stack pointer as `uint64_t` since it's a R4300 register after all, but we do access the contents of the stack while scheduling, since we know how the top of the stack looks like (it's a `regblock_t*`). 

Notice that we also have a `void* stack` variable in `thread_t` which is the bottom of the stack (the beginning of the memory area). Maybe we could have `void* stack_start` and `regblock_t *stack_cur`?

Reply:
If we are accessing other things from the top of the stack, we better have the regblock sruct, it'll be clearer for me when I delve more into the implementation probably. Let me think of a name after fully understanding it 🙂